### ISSUE TITLE
[1.9.0-rc1] docker push performance regression 

### ISSUE THREAD

**Issue by burke** [docker/docker#17038]
Pushing to an S3-backed private registry takes much longer on 1.9.0-rc1 than on 1.8.2, even though pushing to a local registry is substantially faster. When attached to a TTY, most of the time is spent after the "Pushing" progress bar hits completion, but before printing "Pushed". This time is negligible when pushing to a registry on the same machine backed by the local filesystem.

In 1.8.2, we spent *way* more time creating the tarball to send ("Buffering to Disk"), but there was no wait between the upload and "Image successfully pushed".

### Some logs

In the comparisons below:

* Same server is used, before and after the upgrade from 1.8.2 to 1.9.0-rc1
* The top layers, `7191` and `8469` are both almost exactly 400MB
* `registry.shopify.io` is an S3-backed private registry, hosted in EC2 along with the machine we're pushing from here.
* This is actually indicative of a trend on sample sizes > 1 :)

##### Docker 1.8.2

```
Oct 14 18:56:19.277:	The push refers to a repository [registry.shopify.io/shopify-ci-branches] (len: 1)
Oct 14 18:56:19.294:	7191157cb30d: Buffering to Disk
Oct 14 18:56:32.647:	7191157cb30d: Image successfully pushed
Oct 14 18:56:32.928:	fa1c81cb6b28: Image already exists
Oct 14 18:56:33.267:	03c5add71c1d: Image already exists
Oct 14 18:56:33.686:	b25524db876d: Image already exists
Oct 14 18:56:34.024:	c290db7a5070: Image already exists
Oct 14 18:56:34.283:	454c2d886fcc: Image already exists
Oct 14 18:56:35.774:	ba5e0000ba5c: Image already exists
Oct 14 18:56:35.777:	ac192e88bf8434b72d6b237634ebf3bac5ab9ace: digest: sha256:30b3430df4b7d4bbd9f56b678b5a1c35b3632c7fd088eeca86ff93d8070065fe size: 12862
```

##### Docker 1.9.0-rc1

```
Oct 14 19:00:09.679:	The push refers to a repository [registry.shopify.io/shopify-ci-branches] (len: 1)
Oct 14 19:00:09.712:	8469016332f7: Preparing
Oct 14 19:00:09.807:	8469016332f7: Pushing
Oct 14 19:01:36.139:	8469016332f7: Pushed
Oct 14 19:01:36.388:	a5323a34b08b: Image already exists
Oct 14 19:01:36.749:	03c5add71c1d: Image already exists
Oct 14 19:01:37.170:	b25524db876d: Image already exists
Oct 14 19:01:38.101:	c290db7a5070: Image already exists
Oct 14 19:01:38.369:	454c2d886fcc: Image already exists
Oct 14 19:01:38.684:	ba5e0000ba5c: Image already exists
Oct 14 19:01:38.687:	23168f9ff831f4097bed8cb6aeab38e3db9ea3c7: digest: sha256:17092b886b66afb5ec9f66c2df760533dc43c4fd13fd99a1238f51a88f307caf size: 11567
```

**Comment by burke** [comment]
```
burke@locutus4.ec2:~ $ docker version
Client:
 Version:      1.9.0-rc1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   9291a0e
 Built:        Wed Oct 14 03:04:22 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-rc1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   9291a0e
 Built:        Wed Oct 14 03:04:22 UTC 2015
 OS/Arch:      linux/amd64
burke@locutus4.ec2:~ $ docker info
Containers: 1
Images: 134
Engine Version: 1.9.0-rc1
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.19.0-28-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 16
Total Memory: 29.44 GiB
Name: locutus4.ec2.shopify.com
ID: 7R4J:UVLA:UOUE:OZNW:QAZK:QMED:MRF7:LI2L:UWFS:J4H2:WE6V:HSMU
WARNING: No swap limit support
burke@locutus4.ec2:~ $ uname -a
Linux locutus4.ec2.shopify.com 3.19.0-28-generic #30~14.04.1-Ubuntu SMP Tue Sep 1 09:32:55 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
burke@locutus4.ec2:~ $
```

**Comment by burke** [comment]
I'm kind of poking at this myself, will see how far I get this afternoon. Looks at first glance like it's time spent in https://github.com/docker/docker/blob/master/graph/push_v2.go#L263-L266

**Comment by burke** [comment]
Nope, that's not true, here's what I've got:

```
[t=0] about to io.Copy (push_v2.go#L257-L261)
[t=41] done io.Copy. about to digest (push_v2.go#L263-L266)
[t=41] enter *httpBlobUpload Commit
[t=82] enter *blobStatter.Stat
[t=84] exit *blobStatter.Stat
[t=84] exit *httpBlobUpload Commit
[t=84] done digest
```

Previously the whole process took 11 seconds.

**Comment by icecrime** [comment]
Ping @aaronlehmann!

**Comment by aaronlehmann** [comment]
I think this is related to https://github.com/docker/docker/pull/15493

One thing I notice looking at the code is that 1.8.x used `httpBlobUpload.ReadFrom`, which would construct a PATCH request that read from a reader. But the code in 1.9 uses `io.Copy`, which invokes `httpBlobUpload.Write` repeatedly instead. Since `Copy` uses a 32kb buffer internally, this can result in many PATCH requests. I can see why that would be very slow with a S3 backend.

I will work on a fix for this. @burke, if you have time, could you please try reverting `349aef06d51a9425f70a877585fe45283366c2c4` and see if that makes a difference? It reverts cleanly except for a conflict in an integration test. You can just remove that integration test.

cc @stevvooe

**Comment by aaronlehmann** [comment]
I missed how `io.Copy` will use a `ReadFrom` method if available - so ignore the comment above. I'm continuing to investigate.

**Comment by aaronlehmann** [comment]
@burke: I think I understand what's going on here.

The upload to the registry is a two-step process. First the actual data is uploaded; then the engine asks the registry to commit that data to the blob store. The commit phase is where you see the pause.

When the registry uses S3 as the backend, committing to the blob store involves copying the data. This takes a significant amount of time. However, if the blob already exists on the registry, the copy is skipped.

I think the pushes you are doing with 1.8.2 upload layers that already exist, so they can skip the copy. But 1.9.0 is generating blobs with different hashes from those that existed previously, so they need to be copied to the right place.

I tested with 1.9.0 and 1.8.2, and with both versions I see pauses for some time at the end of the "pushing" step. I believe what you're seeing is not a bug, but an artifact of the way you're testing, where 1.8.2 happens to be uploading layers that already exist, and 1.9.0 is uploading layers which involve more work for the registry.

**Comment by burke** [comment]
I don't think that's it. The image below shows push times by host. The last two were upgraded to 1.9.0 yesterday, and it would be exceedingly rare for an image in our hierachies to reference an existing upper layer that hadn't been referenced in the past 5-10 builds (i.e. any effect from having to write out new caches is long-since gone in this data):

![data](https://d1ro8r1rbfn3jf.cloudfront.net/ms_41582/gMOuF7pnV4XFDvkDzJV6OmsW4Oa6zJ/Metric%2BExplorer%2B%257C%2BDatadog%2B2015-10-15%2B14-28-41.jpg?Expires=1445020140&Signature=T2IlZCDm1kF8Yp7YC6~ghmkjtIuogDdtNwa~QkbjCJrsc5yxZcHNNacFDxDzu6ugSci1366Mv~1mde3P4IQdoAzmE-Ixgf-dhb4wH0oc-iHTnE4jJjA-IrPw1TmQXkjHr2D0CEyV43Pcoa3fslgc6GNW1ozpWYKjopEGD~obr2k7xFsHTa2IzyO-pijwCtuSFa1pi1RSnOjsHtcbTQGDvpiYJdbfozFf3vKPuaxb2iNGNkheGRUGiHHDb3T3mUHHHD9-UTAXp4n88lB-SNKhoZTcpdDGJ~9OE8Td3SBJsLAhakUghCEwZDEBNwi4QMIckGBIjNRQO7UoYZvFg27mPQ__&Key-Pair-Id=APKAJHEJJBIZWFB73RSA)

**Comment by stevvooe** [comment]
@aaronlehmann I haven't had a chance to look at this in detail but you are probably on to something with `ReadFrom` getting hidden. `io.Copy` will use it, but if one wraps the reader and doesn't have a `ReadFrom` fallthrough, then it won't be used. We've had similar regressions in the registry that were hit when a Reader didn't have a ReadFrom implementation.

**Comment by aaronlehmann** [comment]
@burke: Thanks for the quick response.

Could you please elaborate a bit on what the workflow looks like? I'm having some trouble imagining why this situation would occur.

For each layer, the engine keeps track of the latest digest it has seen when pushing or pulling it. Before pushing a layer, it checks if the registry has a layer with that digest, and should avoid pushing it at all if the registry already has it.

So unless something is wrong there, the engine should only be pushing newly-built layers. And assuming these new layers are unique, those pushes should involve the overhead of the registry copying them to the final location in the blob store.

**Comment by aaronlehmann** [comment]
@stevvooe: I did some debugging and confirmed that `ReadFrom` was being used. There is only one PATCH request for each blob. Too bad, because that explanation would have made a lot of sense.

**Comment by burke** [comment]
The logs do seem to indicate that both 1.8.2 and 1.9.0 are only pushing the new top layer of our images:

A recent 1.8.2 log: 

```
Oct 15 18:21:45.996:	The push refers to a repository [registry.shopify.io/shopify-ci-branches] (len: 1)
Oct 15 18:21:46.015:	c96f70e4e306: Buffering to Disk
Oct 15 18:21:58.008:	c96f70e4e306: Image successfully pushed
Oct 15 18:21:58.478:	21615e90a830: Image already exists
Oct 15 18:21:58.771:	13fe0d421dcd: Image already exists
Oct 15 18:21:59.185:	b25524db876d: Image already exists
Oct 15 18:21:59.485:	c290db7a5070: Image already exists
Oct 15 18:22:00.021:	454c2d886fcc: Image already exists
Oct 15 18:22:00.427:	ba5e0000ba5c: Image already exists
Oct 15 18:22:00.430:	5f2a2bc0578ae071185203e39a47e1bd1fb39fe2: digest: sha256:e627abba81e9c9f577b1081ce9c44cd5a5e01836d3e33bd2704b5ab706167f14 size: 12862
```

A recent 1.9.0 log:

```
Oct 15 18:31:05.553:	The push refers to a repository [registry.shopify.io/shopify-ci-branches] (len: 1)
Oct 15 18:31:05.568:	9998d416aa37: Preparing
Oct 15 18:31:05.655:	9998d416aa37: Pushing
Oct 15 18:32:19.932:	9998d416aa37: Pushed
Oct 15 18:32:20.445:	50c6998536b6: Image already exists
Oct 15 18:32:20.802:	8db39f7a1862: Image already exists
Oct 15 18:32:21.047:	b25524db876d: Image already exists
Oct 15 18:32:21.456:	c290db7a5070: Image already exists
Oct 15 18:32:21.832:	454c2d886fcc: Image already exists
Oct 15 18:32:22.117:	ba5e0000ba5c: Image already exists
Oct 15 18:32:22.139:	06afc66ec7e9182d90aaeb45da93f45a5a4d8cb7: digest: sha256:07a8798a8491088dabe5063853a001b43f2f8c4fa5700a7bb7224406dd33a141 size: 11596
```

Our workflow is essentially:

* The bottom four layers (`ba5e0000 .. b255`) constitute a baseimage that changes pretty rarely (weeks, sometimes months) and is always shared between build nodes.
* the `8db3` and `13fe` layers are the result of a fixed setup script in the source repo that changes very infrequently. It's usually the baseimage changing that causes this to rebuild. This layer is currently built differently on each build node, hence the different IDs.
* the `50c6` and `2161` layers are the result of checking out some fixed point in the recent past (typically somewhere in the middle of the previous night) and running some things like `bundle install` in that context. This layer is generally shared by all images built later that day.
* The top layer, which is the only one being pushed here in both versions (as expected) is the result of copying in the current source and finalizing compilation.

TL;DR: the bottom 6 layers only change once a day. It looks to me like docker is correctly identifying that only the top layer needs to be pushed, but for some reason it takes much longer on 1.9.0. this layer is 400MB: it includes a `/var/lib/mysql` :disappointed:

**Comment by burke** [comment]
Is it possible that 1.8.2 would just shoot the blob up to the registry and return, but 1.9.0 is waiting for the data to become available via S3 or something like that?

**Comment by aaronlehmann** [comment]
@burke: Thanks for the info, this is very helpful.

Both 1.9.0 and 1.8.2 call the `Commit` method, which is where you are seeing lots of time spent per https://github.com/docker/docker/issues/17038#issuecomment-148176373. This method sends a very small PUT request to the registry, and waits for it to complete. The time is spent waiting for the registry, not doing anything inside the engine. The only explanation that makes sense for major differences in the time spent waiting here is the registry already having the blob in its blob store in one case but not the other. However, I'm at a loss as to why the Docker engine version would make a difference.

If you have access to the registry logs, we can dive into this from the registry's perspective. For each pushed layer that the registry didn't already have, you should see a line like this:

    time="2015-10-15T17:52:07.731240035Z" level=debug msg="s3.Move(\"/docker/registry/v2/repositories/ubuntu/_uploads/db9528b7-bbcf-4bfb-a3b3-04ac0000a7aa/data\", \"/docker/registry/v2/blobs/sha256/7c/7ccf5baddc671df300711753fa329a2b6feb190ea523b019c634920ad66927cc/data\"" environment=development http.request.host="registry:5000" http.request.id=135ab698-9edd-4254-9a60-a6cbde3dc4ab http.request.method=PUT http.request.remoteaddr="162.245.20.130:27790" http.request.uri="/v2/ubuntu/blobs/uploads/db9528b7-bbcf-4bfb-a3b3-04ac0000a7aa?_state=24p2zYtFZBpOe6350AwOlaIdB2XgGwQZqorUniclJDt7Ik5hbWUiOiJ1YnVudHUiLCJVVUlEIjoiZGI5NTI4YjctYmJjZi00YmZiLWEzYjMtMDRhYzAwMDBhN2FhIiwiT2Zmc2V0IjoxOTcyNDkwMjQsIlN0YXJ0ZWRBdCI6IjIwMTUtMTAtMTVUMTc6NTE6MjNaIn0%3D&digest=sha256%3A7ccf5baddc671df300711753fa329a2b6feb190ea523b019c634920ad66927cc" http.request.useragent="docker/1.9.0-dev go/go1.4.3 git-commit/9216b08-dirty kernel/4.0.7-boot2docker os/linux arch/amd64" instance.id=108f00d9-ef8f-4abf-b4eb-bc6ca9f75d22 service=registry trace.duration=11.251662362s trace.file="/home/ubuntu/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Move" trace.id=85165849-d44a-4830-a147-55fc0e180246 trace.line=135 vars.name=ubuntu vars.uuid=db9528b7-bbcf-4bfb-a3b3-04ac0000a7aa version=v2.1.1-54-g3fa64a9.m

Note the "trace.duration" key (11 seconds in this case).

My hypothesis was that on pushes from 1.8.2 (which apparently don't pause between the upload and "Image successfully pushed"), you won't see such a line in the logs. But if it does indeed appear, it would be interesting to know if it shows a consistently shorter trace.duration than an equivalent push with 1.9.0.

**Comment by thaJeztah** [comment]
ping @burke are you able to provide these logs?

**Comment by burke** [comment]
Yes, I'll get those tomorrow morning. Sorry, I got derailed toward the end of last week because we had Hack Days.

**Comment by thaJeztah** [comment]
@burke no problem, I was just browsing through open issues <3

**Comment by aaronlehmann** [comment]
It turns out that there was a regression which means the layers are not getting compressed at all. That is probably why you're seeing slower pushes in 1.9.0-rc1. I've opened PR #17227 to fix this.

**Comment by thaJeztah** [comment]
Thanks @aaronlehmann!

**Comment by burke** [comment]
1.8.2: 16s
1.9.0-rc1: 89s
1.9.0-rc1 + 4dce280d: 47s

**Comment by tiborvass** [comment]
@burke is there a ready-to-use script you could provide us to save some time?

**Comment by burke** [comment]
I'll try to put something together tomorrow.

**Comment by tiborvass** [comment]
@aaronlehmann @tonistiigi @LK4D4 if you guys could help with profiling this that'd be dope :)

**Comment by tonistiigi** [comment]
@tiborvass Isn't this a dupe of #17227

**Comment by tiborvass** [comment]
@tonistiigi #17227 is supposed to fix this. It did help a lot according to https://github.com/docker/docker/issues/17038#issuecomment-150331384 (almost 50% faster) but nowhere near `1.8` according to burke.

**Comment by LK4D4** [comment]
@aaronlehmann @tonistiigi Probably you know better. in 1.8.2 99% of push is compression(probably on disk), however in 1.9 55% takes `net/http/internal.(*chunkedWriter).Write` function.

**Comment by aaronlehmann** [comment]
I haven't had a chance to look yet, but that sounds consistent with smaller chunks of data getting returned from the reader that's passed to `httpBlobWriter`'s HTTP request. Each one would cause a separate HTTP-level chunk, which adds overhead. It makes sense that the gzip compressor would write data to its output in smaller increments than what we would get reading directly from a file.

I'll try buffering the compressor output and seeing if that helps.

**Comment by stevvooe** [comment]
@LK4D4 When I looked at the issues with 1.8.2, most of that "compression" was spent buffering to disk.

**Comment by aaronlehmann** [comment]
The HTTP chunk sizes were indeed tiny. PR #17291 should fix it.

**Comment by burke** [comment]
Confirmed #17291 fixes the problem entirely :sparkles:.

**Comment by thaJeztah** [comment]
Thanks @burke, and thanks for fixing @aaronlehmann, great job!
